src/tool_operate.c

```diff
@@ -697,45 +697,74 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
   return result;
 }
 
+/* helper function to ensure input ends in val_to_ensure */
+static CURLcode ensure_trailing(char **input, const char val_to_ensure)
+{
+  if(*input && **input) {
+    size_t len = strlen(*input);
+    if(((*input)[len - 1] != val_to_ensure)) {
+      struct curlx_dynbuf dyn;
+      curlx_dyn_init(&dyn, len + 2);
+
+      if(curlx_dyn_addn(&dyn, *input, len)) {
+        Curl_safefree(*input);
+        return CURLE_OUT_OF_MEMORY;
+      }
+
+      Curl_safefree(*input);
+
+      if(curlx_dyn_addn(&dyn, &val_to_ensure, 1))
+        return CURLE_OUT_OF_MEMORY;
+
+      *input = curlx_dyn_ptr(&dyn);
+    }
+  }
+
+  return CURLE_OK;
+}
+
 static char *ipfs_gateway(void)
 {
   char *gateway = NULL;
   char *ipfs_path = NULL;
   char *gateway_composed_file_path = NULL;
   FILE *gateway_file = NULL;
 
-  gateway = getenv("IPFS_GATEWAY");
+  gateway = curlx_getenv("IPFS_GATEWAY");
 
   /* Gateway is found from environment variable. */
-  if(gateway && *gateway) {
-    char *composed_gateway = NULL;
-    bool add_slash = (gateway[strlen(gateway) - 1] != '/');
-    composed_gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
-    if(composed_gateway) {
-      gateway = aprintf("%s", composed_gateway);
-      Curl_safefree(composed_gateway);
+  if(gateway) {
+    if(ensure_trailing(&gateway, '/') != CURLE_OK) {
```

Comment:
We usually go with just `if(ensure_trailing(&gateway, '/')) {` when checking CURLcode functions since we know `CURLE_OK` is zero. That saves us from comparing with `CURLE_OK` in checks. It's not a *must*, but it is a standard pattern in curl code. You could simplify several checks here like this.

Reply:
This + comment below. Cleans it up nicely! Thanx :)

For reference. Some jobs back i was "forced" into writing checks verbose. The reasoning there was some people didn't like to have to pay attention to the single exclamation mark and wanted a policy where the `!` at the beginning of an if-statement wasn't allowed... That - now a habit of sorts - kinda stayed with me. Not that verbose is wrong, but short and still readable is better imho.

I like your style better :)